app/models/facilities_query.rb

```diff
@@ -0,0 +1,178 @@
+# frozen_string_literal: true
+
+class FacilitiesQuery
+  attr_reader :params
+
+  def initialize(params)
+    @params = params
+  end
+
+  def query
+    if valid_location_query?
+      location_query_klass.new(params).run
+    elsif @params[:ids]
+      build_result_set_from_ids(params[:ids]).flatten
+    else
+      BaseFacility.none
+    end
+  end
+
+  # Massages the location_query_klass return value to be boolean
+  def valid_location_query?
+    !location_query_klass.nil?
+  end
+
+  # When given more than one type of distance query param,
+  # return nil because the app will not choose preference for the user.
+  # In the controller, this logic is somewhat duplicated
+  # and will render an error when given multiple params.
+  # In the case that only one of these types is given,
+  # return the class used to make that type of query.
+  def location_query_klass
+    @location_query_klass ||= case location_keys
+                              when [] then nil
+                              when %i[lat long] then RadialQuery
+                              when [:state]     then StateQuery
+                              when [:zip]       then ZipQuery
+                              when [:bbox]      then BoundingBoxQuery
+                              end
+  end
+
+  def location_keys
+    (%i[lat long state zip bbox] & params.keys.map(&:to_sym)).sort
```

Comment:
Just a note to say I'm excited to have learned about this `&` set intersection syntax. Ruby can be so terse.

Reply:
ruby set operators are one of my favorite things for some reason